OpenRA.Mods.Common/Traits/Explodes.cs

```diff
@@ -81,5 +90,14 @@ WeaponInfo ChooseWeaponForExplosion(Actor self)
 			var useFullExplosion = self.World.SharedRandom.Next(100) <= info.LoadedChance;
 			return (shouldExplode && useFullExplosion) ? info.WeaponInfo : info.EmptyWeaponInfo;
 		}
+
+		public void Damaged(Actor self, AttackInfo e)
+		{
+			if (info.DamageThreshold == 0)
+				return;
+
+			if ((health.HP * 100) < (100 - info.DamageThreshold) * health.MaxHP)
+				self.Kill(e.Attacker);
```

Comment:
> The remaining issues now are that an enemy scores 2 kills on a fake building while you receive 2 deaths and buildings are demolished with a stronger screen shake than all other buildings.

These happen because of the way that we detect how weapons have killed a target (turns out that our damage handling code is not reentrant).  The short version is: replace this line with `self.World.AddFrameEndTask(w => self.Kill(e.Attacker));` to fix the quoted problems.